### ISSUE TITLE
setup_requires=["numpy"] fails on Windows+Py3.8

### ISSUE THREAD

**Issue by anntzer** [numpy/numpy#14873]
Packages that use setup_requires=["numpy"] in their setup.py appear to fail to build on Windows+Py3.8.

### Reproducing code example:

https://github.com/anntzer/setup_requires-numpy is a minimal repro with just the following setup.py:
```
from setuptools import setup

setup(
    name="testpkg",
    setup_requires=["numpy"],
)
```
and is "built" (`python setup.py build`) on Azure on {linux,mac,windows}+{py3.7,py3.8}.

The resulting build is at https://dev.azure.com/anntzer/setup_requires-numpy/_build/results?buildId=158
As one can see, all builds succeed except windows+Py3.8, which fails with
```
Traceback (most recent call last):
  File "setup.py", line 3, in <module>
    setup(
  File "C:\hostedtoolcache\windows\Python\3.8.0\x64\lib\site-packages\setuptools\__init__.py", line 144, in setup
    _install_setup_requires(attrs)
  File "C:\hostedtoolcache\windows\Python\3.8.0\x64\lib\site-packages\setuptools\__init__.py", line 139, in _install_setup_requires
    dist.fetch_build_eggs(dist.setup_requires)
…distutils.errors.DistutilsError: Setup script exited with error: SandboxViolation: mkd…sers\\VSSADM~1\\AppData\\Local\\Temp\\easy_install-0gx4h9jy\\numpy-1.17.3\\build', 511) {}

The package setup script has attempted to modify files on your system
that are not within the EasyInstall build area, and has been aborted.

This package cannot be safely installed by EasyInstall, and may not
support alternate installation locations even if you run its setup
script by hand.  Please inform the package's author and the EasyInstall
maintainers to find out if a fix or workaround is available.
```

### Error message:

see above

### Numpy/Python version information:

numpy 1.17.3, Python 3.8

**Comment by eric-wieser** [comment]
Possibly a duplicate of gh-3160?

**Comment by anntzer** [comment]
#3160 is against a version of setuptools/easyinstall more than 6y old and a different OS (and setup-requiring numpy has worked fine since then), so I doubt(?) the root issue is the same.

**Comment by eric-wieser** [comment]
Error is truncated above, full error string is
```
distutils.errors.DistutilsError: Setup script exited with error: SandboxViolation: mkdir('C:\\Users\\VSSADM~1\\AppData\\Local\\Temp\\easy_install-0gx4h9jy\\numpy-1.17.3\\build', 511) {}
```
which makes it a little clearer what exactly is escaping the sandbox

**Comment by eric-wieser** [comment]
Looks like the cause here is attempting to write `__config__.py` into a forbidden directory as part of the build - although strange that it only affects 3.8.

**Comment by anntzer** [comment]
Yes, it's definitely the "only Py3.8, only Windows" part that's puzzling me...

**Comment by cgohlke** [comment]
I have not been able to reproduce this issue yet, but I think this could be due to a change in Python 3.8 for Windows that symbolic links and junctions are now resolved by [os.path.realpath](https://docs.python.org/3/library/os.path.html#os.path.realpath). "Symbolic links and junctions" include 8.3 short names, mounted network drives, and substed drives. While `os.path.realpath('path/to/numpy-1.17.3/setup.py')` will be resolved correctly, `os.path.realpath('path/to/numpy-1.17.3/build')` might not because the build directory does not yet exist when the setuptools sandbox compares path names.

Btw, this change in Python 3.8 also prevents running the numpy and scipy tests when `os.path.realpath(sys.executable) != sys.executable`

There are a few other mentions of `SandboxViolation` exceptions on Python 3.8 that could be the same issue, e.g. https://github.com/pypa/setuptools/issues/1874 and https://superuser.com/questions/1496720/cant-install-anything-on-pip

**Comment by Ranothil** [comment]
Also having this issue

**Comment by mattip** [comment]
I do not think this is a numpy-specific problem, but reflects on the use of `setup_requires` in general.

I think it would be better to replace it with a `requirements.txt` or a `pyproject.toml` file. Use of `setup_requires` will force compilation of numpy, which is probably not the desired behaviour. pytest-dev/pytest-xdist#136 is a discussion of this problem, a work-around hack, and the conclusion that the hack is fragile.

**Comment by anntzer** [comment]
This has been fixed in setuptools 42.0.0, which switched to using pip to resolve setup_requires (https://setuptools.readthedocs.io/en/latest/history.html#v42-0-0); see build at https://dev.azure.com/anntzer/setup_requires-numpy/_build/results?buildId=159.
I think this issue can be closed unless you want to support this specific use case with older setuptools.

**Comment by mattip** [comment]
Thanks for getting back with a fix. I will close this, hopefully the search engines will let people find the setup_requires in this issue and figure out the solution is to update setuptools.
